### ISSUE TITLE
Cannot upload to UMA storage buffer without an unnecessary copy and unnecessary memory use

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#2388]
WebGPU currently has 2 buffer upload facilities: `GPUBuffer.mapAsync()` and `GPUBuffer.writeBuffer()`.

For `GPUBuffer.mapAsync()`, there currently is the restriction that no mappable buffer can be used as anything else, other than COPY. This means that, in order to be useful, an application has to allocate 2 buffers - one for mapping and one for using. And, if an application wants to round-trip data through a shader, they have to allocate 3 buffers - one for the upload, one for the download, and one for the shader. Therefore, in order to use `mapSync()`, an application needs to double (or triple) their memory use and add one or two extra copy operation. On a UMA system, neither the extra allocation nor the copy is necessary, which means there's both a perf and memory cost to using `mapAsync()` on those systems. What's more, because the application is explicitly writing this code, there's not really anything we can do to optimize out the extra buffer allocation / copy operation.

On the other hand, `GPUQueue.writeBuffer()` is associated with a particular point in the queue's timeline, and therefore can be called even when the destination buffer is in-use by the GPU. This means that the implementation of `writeBuffer()` is _required_ to copy the data to an intermediate invisible buffer under-the-hood, even on UMA systems, and then schedule a copy operation on the queue to copy the data from the intermediate buffer to the final destination. This extra allocation and extra copy operation don't necessarily need to exist on UMA systems. (`GPUBuffer.writeBuffer()` is a good API in general because of its simple semantics and ease of use, but it does have this drawback.)

It would be valuable if we could combine the best parts of `GPUBuffer.mapAsync()` and `GPUQueue.writeBuffer()` into something which doesn't require an extra allocation or copy on UMA systems. This kind of combination would have to be something that isn't UMA-specific, but would work on both UMA and non-UMA, and UMA systems would be able to avoid extra allocations/copies under-the-hood.

1. The "async" part of `GPUBuffer.mapAsync()` would be valuable, because that allows the implementation to not have to stash any data due to the destination buffer being busy.
2. The "map" part of `GPUBuffer.mapAsync()` would be valuable because it allows the array buffer to be backed directly by GPU memory, thereby potentially avoiding another copy on UMA systems.
3. The "queue" part of `GPUQueue.writeBuffer()` would be valuable, because non-UMA systems _would_ need to schedule an internal copy to the destination, and specifying the queue gives them a place to do that.

I think the most natural solution to this would be:
1. Give `mapAsync()` an extra `GPUQueue` argument. (`getMappedRange()` and `unmap()` will implicitly use this queue)
2. Relax the requirement that the only other usage a mappable buffer can have is COPY

- On a UMA system, you'd be able to map the destination (storage) buffer directly - No copies, no extra allocations, it's living the UMA dream.
  - For reading, `mapAsync()` would just ignore its `GPUQueue` argument.
  - For writing, `mapAsync()` would use its `GPUQueue` argument to schedule a `clearBuffer()` command of the relevant region of the buffer. After the clear operation is complete, the map promise would be resolved.
- On a non-UMA system:
  - For reading, `mapAsync()` would schedule a copy from the source (storage) buffer to a temporary buffer using the specified GPUQueue (which is what an application would have had to do themself), and the map operation would proceed just like normal on the temporary buffer. This is exactly what an author would have had to do themself.
  - For writing, `mapAsync()` would just stash the queue, map a temporary buffer, and wait for `unmap()` to be called. When `unmap()` is called, it would schedule a copy on the stashed queue from the temporary buffer to the destination buffer. This is exactly what an author would have had to do themself.

It's important to note is that this proposal doesn't restrict the amount of control a WebGPU author has. If an author wants to allocate their own map/copy buffer and copy the data to/from it on its way to its final destination (as they would do today), they can still do that, and no invisible-under-the-hood temporary buffers would be allocated.

**Comment by Kangz** [comment]
Overall I'm worried about modifying the buffer mapping mechanisms this late, when it was perhaps the single most difficult part of the API to find a good design for and reach consensus on. I think a UMA optional feature would make a 

We discussed the slight inefficiencies that UMA has with the current mapping mechanism multiple times in F2F, and https://github.com/gpuweb/gpuweb/issues/605 suggests a UMA feature could be done to optimize this later (later could be now). The proposal is interesting but has some issues.

Side note: there is also `mappedAtCreation` which was added so that the initial upload of data in buffers can be made perfectly efficient on UMA systems. (up to the copies necessary for the process separation).

### Multiple buffers per `GPUBuffer`

It breaks down the model that one WebGPU buffer == one underlying API buffer. This is a pretty useful thing to keep because it makes it very clear to the developer what the memory cost of things is. That you can have temporary staging for `mappedAtCreation`, and potentially shmem wrapped in the `ArrayBuffer` given to JS is already very difficult for developers to reason about in terms of cost.

### Cost of consistency for `MAP_WRITE`

Currently `MAP_WRITE` buffers give you an `ArrayBuffer` that contains the current content of the buffer. Since the buffer can only be written by Javascript, no copies are ever needed to update the content of the buffer, it's just the `ArrayBuffer` wrapping shmem shared between the GPU and Web processes. If the GPU can write to the buffer, then we need to copy data from the UMA buffer to the shmem (or even worse, from VRAM to readback to shmem).

On the other side, if we say that `mapAsync(MAP_WRITE)` always zeroes the buffer, then in most cases the CPU has to zero the buffer, since the UMA/readback buffer isn't shared with the Web process. Either there's a `memset(0)` or a `memcpy` from the UMA buffer to the shmem.

### Consistency for `MAP_READ`

What happens when Javascript writes into the buffer mapped for reading? Assuming you are able to create an `MTLBuffer` from a shmem FD to reduce the number of copies to the maximum, the writes that Javascript did all of a sudden become visible to the GPU, while an all other configurations, Javascript writing to the buffer doesn't have any visible effect for the GPU?

### Relatively little gains and a feature proposal

The gains you get with the proposal you suggested seem small: if you have a large amount of data to initialize buffers with, then you can use `mappedAtCreation` that's the optimal path. If you need to modify part of a buffer while it's in use, then you have to schedule a copy because mapping is an ownership transfer of the full buffer (I tried to figure out how to do sub-range mapping efficiently but gave up).

So the cases this help are when you need to upload data to a buffer that's not being created, but also not currently in use by the GPU. This should be a fraction of the actual buffer transfers. Still might be worth speccing an optional feature, but not modifying the core buffer mapping spec.

The optional "UMA" feature could:
 - Lift the restriction for `MAP_WRITE` to allow any other read-only usages.
 - `MAP_READ` already allows all the write-only usages. But maybe more are added in the future so the extension would also lift that? Or it allows it with any other usages, but assumes there is always a UMA -> shmem copy happening in the GPU process (so that JS writes are never made visible to the GPU).

**Comment by kainino0x** [comment]
I was going to have comments but @Kangz covered everything I was going to say and more.

**Comment by litherum** [comment]
> It breaks down the model that one WebGPU buffer == one underlying API buffer.

This isn't true. This proposal requires scratch space, certainly, but so does `writeBuffer()`. It's no worse.

**Comment by litherum** [comment]
> What happens when Javascript writes into the buffer mapped for reading?

This is a good point! I suppose this proposal only makes sense for read/write buffers (which we don't have today, but I think has a natural path forward).

**Comment by kainino0x** [comment]
It's relatively rare for an application to actually need read/write mapping. Sure we could add them for this use-case, but applications would still need to know which one to use and explicitly switch between them based on whether the adapter is UMA or not.

**Comment by Kangz** [comment]
> This isn't true. This proposal requires scratch space, certainly, but so does writeBuffer(). It's no worse.

`WriteBuffer` is quite explicitly an implementation-managed ringbuffer. But it is not tied to a `GPUBuffer`, it's only `GPUDevice` extra memory. Plus it doesn't need to be persistent data. The implementation can destroy the ringbuffers when there is memory pressure while extra backings for `GPUBuffer` would have to stay, otherwise you could get an OOM trying to map the buffer.

**Comment by benvanik** [comment]
I'm not saying I'm for or against any proposal here, only voicing that I agree what's in the API today does not fully satisfy workflows with dynamic data moving across host/device and that this will be a performance issue in real-world usage. In compute workloads getting data back from the device is a major part of the upload -> compute -> download flow and until we have a `GPUQueue.readBuffer` (🙏 please!) this results in non-trivial complexity in user code and bloat.

> But it is not tied to a GPUBuffer, it's only GPUDevice extra memory.

👍  IMO having the implementation manage the ringbuffer with writeBuffer/readBuffer and incurring a copy is acceptable if the alternative is managing exclusively upload or download buffers in user code (as I found the spec detail that indicates a buffer cannot be both, resulting in user staging pools needing double the memory for bidi transfer). This way multiple libraries trying to perform the upload/download are not each keeping around large GPUBuffers for this purpose.

**Comment by litherum** [comment]
> you could get an OOM trying to map the buffer.

This is no worse than the possibility of an OOM when trying to `writeBuffer()`, though...

**Comment by Kangz** [comment]
> Sure that's a possibility as well, although implementation could stall to free staging space if they really want to.

The point here wasn't that `writeBuffer` can or cannot OOM, it is that you want to give developers a way to do transfers without the possibility to trigger unfixable OOM. Buffer mapping can do that since OOM only happens at buffer creation. If you choose to make staging / readback buffer transient, then you lose this control in the application because you can OOM on `mapAsync` as well. It is possible to decide to do that, but need to be cognizant of all the tradeoffs we are making. In this whole comment thread I suggest it is a bad idea for many reasons, including that OOM issue.

**Comment by kvark** [comment]
I agree with the concerns about managing temporary buffers for mapping, expressed by @Kangz . Their lifetime is attached to mapping, and it's worse than the ring buffer we currently have for the writes.

I also agree with @litherum that it would be good to be able to avoid copies on systems that can do that.
An optional feature for UMA architectures seems like the right way to proceed. It would basically lift the restriction on usages for buffers, allowing MAP_READ+MAP_WRITE+anything else.

As for the queue argument for mapping, this correlates with https://github.com/gpuweb/gpuweb/issues/1977#issuecomment-884436273. It's probably needed.

**Comment by Kangz** [comment]
I'm happy to help by writing an optional feature for UMA that allows `MAP_READ + WRITE`.

**Comment by litherum** [comment]
It would be pretty unfortunate if authors had to opt-in to avoid using 2x memory on UMA machines.

**Comment by kainino0x** [comment]
I don't think anyone is disagreeing about that, but if we're going to avoid it we're going to need a proposal that works. I don't think we're getting any closer to one.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-02-23#cannot-uploaddownload-to-uma-storage-buffer-without-an-unnecessary-copy-and-unnecessary-memory-use-2388">WGSL meeting minutes 2022-02-23</a></summary>



* KN: nobody satisfied with current state, but nobody has a better idea. Everyone's resigned to this fate except Myles. :)
* KG: one thing that has changed since it was first discussed - more common today than 2 years ago to get adapters that let you map CPU read and host read/device use - used to be UMA archs only, and some AMD cards - but has changed now.
* KN: right. Intel doesn't even have some of these options (host-coherent + device-coherent?). Think we could do this on Intel regardless.
* KG: if something we can't support - don't want to fragment the ecosystem by making you write 2 paths. If things have changed - still do need that.
* KN: we don't have solution for doing this underneath the hood of the application. Can do it with an extension. Would like to. Maybe we should do it for 1.0. Would need separate code path for application.
* KR: WebGL doesn't have the ability to optimize for this and performance in this area is basically fine. I think WebGPU will also perform fine in general without this optimization, and since applications will have to add a new code path to take advantage of it, think this should be pushed out to post-V1.


</details>

**Comment by Kangz** [comment]
So here's the proposal for the extension: what I wrote above

```
The optional "UMA" feature could:

 - Lift the restriction for MAP_WRITE to allow any other read-only usages.
 - MAP_READ already allows all the write-only usages. But maybe more are added in the future so the extension would also lift that? Or it allows it with any other usages, but assumes there is always a UMA -> shmem copy happening in the GPU process (so that JS writes are never made visible to the GPU).
```

With the addition that if readonly `ArrayBuffers` become a thing, then we can lift all restrictions on `MAP_READ` (except `MAP_WRITE`? not sure), by making the `ArrayBuffer` returned by mappings for reading be readonly.

**Comment by Kangz** [comment]
As discussed in the meeting, moving to post-V1 polish since the only proposal so far is an optional feature.

**Comment by ErichDonGubler** [comment]
@Kangz: You appear to have a broken sentence [here](https://github.com/gpuweb/gpuweb/issues/2388#issuecomment-988144669):

> I think a UMA optional feature would make a

**Comment by kainino0x** [comment]
> @Kangz: You appear to have a broken sentence [here](https://github.com/gpuweb/gpuweb/issues/2388#issuecomment-988144669):
> 
> > I think a UMA optional feature would make a

I think the gist was "UMA would make sense to put in an optional feature"

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2023-06-07-08-(Pacific-time)-Meeting-Minutes#socialization-cannot-uploaddownload-to-uma-storage-buffer-without-an-unnecessary-copy-and-unnecessary-memory-use-2388">GPU Web 2023-06-07/08 (Pacific time)</a></summary>

* Recap the design constraints for this problem
* MM: wanted to touch base before going off and doing a bunch of engineering
    * We're interested in UMA working well
    * Interested in a potential solution where the same code would "do the right thing" on UMA and non-UMA
    * This group posited that that was not possible
    * I think it might be
    * Want to nail down what the original objections were
* KR: from our side we need enga@ and cwallez@ present for the conversation. Would like to advance this on the Github issue or mailing list.
* Postpone for a week?
* KG: I can try to synthesize
* KG: on non-UMA archs you sometimes need 2 copies, and on UMA you can get to 1 copy. How to pipeline, prioritizing bandwidth/latency, is where Corentin and I ran aground trying to find a single API to do both.
* KG: My position - if you try to figure out the API for these things, you'll either prove us wrong or right, and that's great
* MM: that's reassuring. Think we're in a different situation now than 2021. Now we have 2 ways of getting data on the card. I'd be coming back with a 3rd way. Adding a 3rd way isn't great for the platform, but if an app cares about the tradeoffs, we'd have more options for them.
* **Continue this next week**.

</details>

**Comment by MikhailGorobets** [comment]
The current restriction on buffers created with map flags has problems also on NUMA architectures. Small, frequently updated uniform buffers can be stored in system memory without significant impact on performance. In addition, with the advent of Resizable BAR and SAM, it is possible to write data directly to VRAM using the CPU (we can even write textures directly to VRAM and change later the access pattern from linear to swizzled, for better bandwidth)

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-10-F2F#uma-buffer-mapping-2388-albin">GPU Web WG 2024-10-29/30 Mountain View F2F</a></summary>

* Interested: Mike Wyrzykowski
* AB: Issue bigger than just memory bandwidth. Also serious performance impact on barriers. (no one was taking notes here). End up with all-graphics–to–all-graphics barrier. 15% regression just from barriers.
* …
* AB: If we know the buffer is not scheduled for use, but it's mappable (in vulkan terms, which is basically all buffers on UMA), and any gpu writes are visible to the host, we know we can skip the upload buffer and write directly. For all of these conditions to be true, there is a hidden fast path where developers have to keep one buffer per frame in flight.
* CW: This is why it gets a star. Being in the browser makes it more complicated…
* JS: Does UMA mean zero-copy? Reduced copies?
* AB: 
* CW: In WebGPU we have this idea we call "triply mapped buffers" which are visible to the GPU, the browser "GPU process", and the browser "content process" all at once. Possible but very experimental. On Windows may stress the memory system of the OS. On Vulkan not super clear how many devices support it and how well. So unfortunately can't guarantee mappable buffers get triply mapped.
* JS: If there's one copy from content process to GPU process and zero-copy after that, is that OK?
* AB: Zero copy ideal, but eliminating synchronization barriers really good. I've been profiling this mostly in native WebGPU so no browser processes involved.
* CW: Overall, yes please: want to avoid the expensive memory barriers. Very interested in having your input, and as an oracle of whether it actually fixes the memory barriers.
* MW: Many devices with low memory limits like 1.5GB for a web process, zero-copy helps a lot with this [reducing memory pressure].
* KG: whiteboard:
    * UMA pool sizes:
        * None: Harsh Reality
        * Partial: uniforms, dynamic vert data
            * On discrete, have about 256MB memory you can DMA from CPU
        * Full: texture data uploads
            * But with ReBAR you can get large DMA regions.
* JS: Addresses barrier issues?
* CW: No. Think we need to collect all of the problems and try to find a solution. It won't be perfect.
* CW: Know in the past we said we didn't want to expose on discrete because it's pessimizing. But maybe not true. OK if the "UMA" API is slower on discrete.
* KG: Need data. If this "None" category is empty, then it becomes a very interesting thing to have some partial UMA with possible performance cliff past 256MB or whatever.
* CW: Not sure how …
* JS: But doesn't solve synchronization bubbles.
* CW: Since our mappable buffers can't be used for anything the user has to enqueue a copy. … applications can have rolling buffers 

</details>

**Comment by Jiawei-Shao** [comment]
Hi all,

I've just submitted the [proposal](https://github.com/gpuweb/gpuweb/pull/5108) on the WebGPU optional feature `buffer-map-extended-usages` that adds the support of mappable buffer with any other buffer usages.

## Motivation

1. Reduce the memory footprint when transfering a large amount of data between CPU and GPU.
2. Eliminate the extra copy between the staging buffer and the destination buffer and the related barriers.

## New Features

1. Allows creating a mappable GPUBuffer (created with `MAP_READ` or `MAP_WRITE` usages) with any other GPUBufferUsage flags, including the use of both `MAP_READ` and `MAP_WRITE`.
2. Allows `READ|WRITE` as a valid value for the `mode` parameter in `GPUBuffer.MapAsync()`.

## Implementation Details

### Platform Requirements

`buffer-map-extended-usages` can be supported efficiently on the backends with UMA architecture and below conditions.

|Backends|Preferred Requirements|
|--------|----------- |
|D3D12   |`D3D12_FEATURE_DATA_ARCHITECTURE.isUMA && `<br>`D3D12_FEATURE_DATA_ARCHITECTURE.isCacheCoherentUMA`|
|Metal   |`[MTLDevice hasUnifiedMemory] == true`|
|Vulkan  |`VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT` &#124;<br> `VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT` &#124;<br> `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` &#124;<br> `VK_MEMORY_PROPERTY_HOST_CACHED_BIT`  |

**Notes**
- **D3D12**<br>
`D3D12_FEATURE_DATA_ARCHITECTURE.isCacheCoherentUMA == true` is recommended according to [this document](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_feature_data_architecture#how-to-use-uma-and-cachecoherentuma).

- **Vulkan**<br>
  - `HOST_VISIBLE_BIT` is required for the access on CPU;<br>
  - `DEVICE_LOCAL_BIT` is preferred for the access on GPU;<br>
  - `HOST_CACHED_BIT` is preferred as such memory type provides cached storage on the CPU;<br>
  - `HOST_COHERENT_BIT` is preferred because without this bit we need to manually call `vkFlushMappedRanges()` when CPU has completed writing data and `vkInvalidateMappedRanges()` when the GPU data has been written back.<br>

  According to [this document](https://developer.arm.com/documentation/101897/0304/CPU-overheads/Vulkan-CPU-memory-mapping), "do not read back data from uncached memory on the CPU", so we require the at least `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_CACHED_BIT` for `buffer-map-extended-usages`, and `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_CACHED_BIT | HOST_COHERENT_BIT` is preferred to **guarantee the performance of CPU read**.

  `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_CACHED_BIT | HOST_COHERENT_BIT` is only available on [39.85%](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=15&platform=all) Vulkan devices with no Mali GPU supported, while the coverage of `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_COHERENT_BIT` is [56.32%](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=11&platform=all).

### Data Transmissions in a Multi-processing Browser

|Map Mode|Buffer Usages|Return values of `GetMappedRange()` | Data in the returned array buffer After `Unmap()` |
|-----------|------------------|--------------------------|---------------- |
|`MAP_READ` |-                 |Buffer's current values|Discarded|
|`MAP_WRITE`| Without `STORAGE`|Buffer's current values|Stored in the GPUBuffer|
|`MAP_WRITE`| With `STORAGE`   |The default initialized data (zeros) or<br>data written by the webpage during a previous mapping|Stored in the GPUBuffer|
|`MAP_READ`&#124;`MAP_WRITE`|-|Buffer's current values|Stored in the GPUBuffer|

**Notes**

The buffer data with `STORAGE` usage may be changed on the GPU side, while the map mode `MAP_WRITE` (without `MAP_READ`) implies that no GPU data will be read back to the returned ArrayBuffer, so under such map mode the content in the returned ArrayBuffer can just be the default initialized data or the data written during a previous mapping.


## Related Topics

### `MAP_WRITE` + any other read-only `GPUBufferUsage` flags in WebGPU Core SPEC

On non-UMA architecture the above combination can also be implemented with an internal staging buffer and an implicit buffer-to-buffer copy inside WebGPU implementation.

- I prefer not supporting it in the WebGPU core SPEC as the staging buffer cannot be managed and even noticed by the developer, while I think WebGPU should give the developer more controls on the GPU resources.

### `MAP_WRITE` + any other `GPUBufferUsage` flags with zero GPU copy as an optional WebGPU feature

The above combination can be supported on the platforms that support `BufferMapExtendedUsages` or the ones with Resizable Base Address Register (`ReBAR`) enabled.

`ReBAR` is a PCIe capability that allows the PCIe device, such as a discrete graphics card, to negotiate the BAR size to optimize system resources. Without `ReBAR` the CPU can only access small portions of the GPU memory (`256MB`) at a time. With `ReBAR` the CPU can access the entire GPU memory (VRAM).

Below are the platform requirements that can explicitly take advantage of `ReBAR`:
|Backends|Requirements|
|--------|----------- |
|D3D12   |Supports heap type `D3D12_HEAP_TYPE_GPU_UPLOAD`:<br>`D3D12_FEATURE_DATA_D3D12_OPTIONS16.GPUUploadHeapSupported == true`<br>(since Windows 11 24H2) |
|Metal   |None|
|Vulkan  |`VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT` &#124;<br> `VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT` &#124;<br> `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` and <br>`VkMemoryHeap.size` is the whole GPU memory instead of `256MB`|

- For the WebGPU implementation with a standalone GPU process it is fine as we always need to copy data from CPU to GPU.
- For the WebGPU implementation with single process there is a problem that the CPU readback is slow as there is no cache on the CPU side for `ReBAR`, so it will be very slow to read any data from that mapped pointer.

**Comment by kainino0x** [comment]
> `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_CACHED_BIT | HOST_COHERENT_BIT`[=15] is only available on [39.85%](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=15&platform=all) Vulkan devices with no Mali GPU supported

> while the coverage of `DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_COHERENT_BIT` is [56.32%](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=11&platform=all).

That link is to [`DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_CACHED_BIT`=11](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=11&platform=all). 56% including 45 AMD, 20 NVIDIA, 43 Intel, 785 ARM, 1234 Qualcomm, 227 ImgTec.

[`DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_COHERENT_BIT`=7](https://vulkan.gpuinfo.org/listdevicescoverage.php?memorytype=7&platform=all) is 85% including 849 AMD, 323 NVIDIA, 151 Intel, 802 ARM, 1168 Qualcomm, 236 ImgTec.

BTW, both have a few extra devices which used extension versions of the same capability bits.

**Comment by JMS55** [comment]
Can you give some examples of how you expect users to use the new APIs? For when users want to support the fastest path for UMA/ReBAR systems, but still also support other systems. I'm having a hard time figuring out what the proposal means in practice.

I'm very excited for this capability though, as without dedicated transfer queues or the proposed changes, there's no way to upload to buffers without blocking the rendering queue.

**Comment by JMS55** [comment]
For _texture_ uploads, we still need something like VK_EXT_host_image_copy to not take up the device queue while uploading textures.

**Comment by sebbbi** [comment]
I don't understand the requirement for HOST_CACHED_BIT. We need a fast zero-copy path to upload data from CPU to GPU. A developer would write data from CPU->GPU, so there's no need for CPU side caching. Write combining will do everything we need. 

Of course programmers have to be careful with write combining to avoid performance pitfalls, but graphics programmers have been used to that since the 20-year-old DirectX and OpenGL with write-only buffer map (those return write-combined memory pointer). 

Mobile GPUs have heaps of this type:
DEVICE_LOCAL_BIT | HOST_VISIBLE_BIT | HOST_COHERENT_BIT

There's no support for HOST_CACHED + HOST_COHERENT on many devices. Could we remove the cached requirement?

**Comment by kainino0x** [comment]
From the [minutes](https://docs.google.com/document/d/1Mn2ODKiEZIBqAY7hFKD8hVCEUAYCbOosYq-KV3CkNaM/edit?usp=sharing) (to be posted here soon):

- CW: problem with just coherent + non-cached is: we're hoping to "triply map" buffers. Create shmem in JS process, send to GPU process, use API-specific method to import that memory into GPU address space. VK_external_memory_host (_fd?). Also Metal and D3D12. If we do this, JS has an ArrayBuffer that becomes uncacheable. Seems weird to expose uncacheable memory to JS. [surprising performance properties]

Of course we could avoid triply-mapping this type of buffer on systems where we can't get HOST_CACHED. As Jiawei's investigation says these bits are only "preferred", not required.

**Comment by sebbbi** [comment]
Yes, the minuses are that you need to understand performance implications for write-combined memory. But my argument is that all low level gfx programmers already understand it, since it's the same on all other APIs, including older ones that didn't support persistently mapped buffers. If WebGPU requires CPU cached + coherent, then that feature is pretty much unusable on Android, since there's not enough coverage.

**Comment by sebbbi** [comment]
At least provide a flag for professional graphics programmers to enable support for non-cached upload heap. I have no problems with that. And most of the professional users will write WASM with C/C++ or Rust anyways, so JS ArrayBuffer is not even visible for us.

**Comment by Kangz** [comment]
This thread form is not super great to discuss detailed feedback on a proposal like this that has a lot of details.

## For `CACHED` and `COHERENT`

@Chainsawkitten explained in the meeting that requiring `CACHED | COHERENT` is not reasonable. The big concern from the Chromium side of things is if we are able to triple-map (or even quad-map: JS, WASM, GPU process and the GPU) uncacheable memory, then for the first time, JS and WASM devs will have to deal with uncached memory. Mapped `GPUBuffers` will result in `ArrayBuffers` that may have orders of magnitudes of perf difference between cached and uncachend memory. IMHO it is a portability concern and we should try to avoid it. I understand that in native it isn't a concern though and that devs rely on write combining to do the correct thing.

My suggestion would be that for triply-mapped buffers we only require `CACHED` and add the necessary invalidation/flush in the implementation. For buffers that require an intermediate copy, the implementation would be free to use `COHERENT` only.

@sebbbi, I didn't find good resources about how to make sure write-combining avoids any uncached reads (though I remember one from years ago). Do you have one? Is it necessary to fully write each cache line touched by a write (to avoid having to load the cache line to combine partial writes)?

## Care about GPU-based validation

We need to make sure that this extension doesn't add any TOCTOU attacks where JS or the Web process is able to modify data in buffers after it has been validated, and also that it doesn't prevent future ways we might want to validate data. At the moment only validation of indirect draw/dispatch requires GPU-side validation that could be subject to TOCTOU, so my suggestion would be to not allow `INDIRECT` with any mappable usage. WebKit also had an idea to validate index buffers on the GPU but I'm not sure about that anymore. @mwyrzykowski are TOCTOU on `INDEX` buffers a concern?

## Multi-process behavior and extra copies

The behaviors listed for the various `GPUMapMode` may not minimize the number of copies depending on whether triply-mapping is happening vs. not. Note that at this point we expect all browsers to ship WebGPU as a multi-process implementation. @mwyrzykowski this is related to your comment on the PR.

With triply mapping:

 - `MAP_READ` needs to make a copy of the contents of the buffer, so as to discard writes.
 - `MAP_WRITE` (GPU read only) no extra copies
 - `MAP_WRITE` (GPU read write) an extra `memset(0)` of the whole mapped range.
 - `MAP_READ | MAP_WRITE` no extra copies

Without triple mapping (one copy always happen between the GPU and JS):

 - `MAP_READ` no extra copies
 - `MAP_WRITE` (GPU read only) one extra copy if the application doesn't care about the previous contents and overwrites all of it.
 - `MAP_WRITE` (GPU read write) no extra copies if the implementation keeps the previous CPU-side contents.
 - `MAP_READ | MAP_WRITE` no extra copies

Also it is a bit weird how the `MAP_WRITE` has different behavior depending on the buffer usage flag. IMHO that's action at a distance and a footgun. Likewise for the behavior of `MAP_WRITE` read-write being one of two possibilities, that's a portability concern.

Some ideas of how we could maybe improve this:

 - Have an explicit `ZERO / DISCARD` flag that can be used by the application to have more defined behavior.
 - Consider zeroing memory with GPU commands on `mapAsync` instead of a memset.
 - Having a hint for the application on whether a buffer is triply mapped. (can we even know this synchronously in Chromium? not sure)

IMHO the ideal solution, if it exists, would have strong portability guarantees across browsers and types of buffers (triply mapped vs not), have clear optimal paths that applications can use and no additional copy overhead than necessary for triply-mapped buffers (while still being very good for non-triply mapped).

## How to use this feature

I'll let @Jiawei-Shao detail. Do you have sample code showing how an application would take advantage of this feature, and write code that can be most efficient with and without it? For @JMS55's question, I believe that this feature would be similar (albeit slightly more limited) than wgpu's [`MAPPABLE_PRIMARY_BUFFERS`](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.MAPPABLE_PRIMARY_BUFFERS).

**Comment by mwyrzykowski** [comment]
> . [@mwyrzykowski](https://github.com/mwyrzykowski) are TOCTOU on `INDEX` buffers a concern?

Yes, it's still a concern. We perform index buffer validation as needed via VS:VS memory barriers on apple graphics chips, falling back to pass splitting on Intel and AMD devices. As long as the copy of memory from JS / WASM is distinct from the UA's memory, then I think we can identify all the places where we must re-validate. 

If the GPU memory was shared with JS, which WebKit also doesn't current support, then it would be much harder to perform such validation via that mechanism.

**Comment by Jiawei-Shao** [comment]
Here are several updates after yesterday's meeting.

### Use of Triply mapping

Below are the possible usages of Triply-mapping on every GPU architectures. As readable mapping requires the performance of CPU read, it is not suitable to use triply mapping on non-CPU-cached UMA.

|                               |CPU-cached UMA|Non-CPU-cached UMA|Resizable BAR|
|-------------------------------|--------------|------------------|-------------|
|MapRead                        |Non-Triply    |Non-Triply        |Not Supported|
|MapRead &#124; MapWrite        |Triply        |Non-Triply        |Not Supported|
|MapWrite                       |Triply        |Non-Triply        |Non-Triply   |


### Behavior of CPU read after map-write-only

We should define the initial data in the array buffer returned by `getMappedRange()` after the write-only buffer mapping.
- With Triply Mapping (on CPU-cached UMA only), the data is exactly the current GPU data.
- Without Triply Mapping, the data may not be the same as the current GPU data.

Shall we always require a clear here or allow keeping the last uploaded data?

**Comment by Jiawei-Shao** [comment]
Now there are two issues about write-only buffer mapping:
1. The performance of CPU read on **the array buffer returned from getMappedRange()**
   The CPU read from non-CPU-cached memory is very slow with triply mapping on non-CPU-cached architectures. 
2. The initial data in **the array buffer returned from getMappedRange()**
   With triply mapping the array buffer exactly contains the data on the GPU, while with non-triply mapping the initial data may not always be the data on the GPU.

So I think the main problem here is that although we are doing write-only buffer mapping we still have to upload data from an array buffer returned from `buffer.getMappedRange()` which can be read but actually we don't care about the initial data of it.

Maybe we can add another API (e.g. `buffer.mapWriteData(arrayBuffer)`) to allow directly uploading data into the write-only mapped buffer without returning the current data on the CPU? The `arrayBuffer` here is managed by the developer and can be reused for other buffers or used for other purposes. `buffer.getMappedRange()` can still be preserved but we can add a new restriction that the data should be cleared to 0 when the buffer has GPU-write usage.

**Comment by mwyrzykowski** [comment]
> . `buffer.mapWriteData(arrayBuffer)`)

Would that be different from `device.queue.writeBuffer(buffer, 0, arrayBuffer)`?

**Comment by Jiawei-Shao** [comment]
`buffer.mapWriteData(arrayBuffer)` can do data uploading at once, while `device.queue.writeBuffer(buffer, 0, arrayBuffer)` can't when buffer is in use, in this time a staging buffer, an extra copy and a barrier are still needed.

**Comment by Kangz** [comment]
A `buffer.mapWriteData(arrayBuffer)` would be interesting but ultimately that's one more copy as well because the application needs to stage the data somewhere and then ask the browser to do the copy.

**Comment by Jiawei-Shao** [comment]
> A `buffer.mapWriteData(arrayBuffer)` would be interesting but ultimately that's one more copy as well because the application needs to stage the data somewhere and then ask the browser to do the copy.

I think one good thing for `buffer.mapWriteData(arrayBuffer)` is that the `arrayBuffer` can be reused just after the call of this API, while with current WebGPU API the mappable staging buffer cannot be reused until the copy is completed on GPU and the `mapAsync()` is resolved.

**Comment by grovesNL** [comment]
> I think one good thing for `buffer.mapWriteData(arrayBuffer)` is that the `arrayBuffer` can be reused just after the call of this API, while with current WebGPU API the mappable staging buffer cannot be reused until the copy is completed on GPU and the `mapAsync()` is resolved.

For what it's worth, [wgpu has a similar (native-only) function called `write_buffer_with`](https://wgpu.rs/doc/wgpu/api/queue/struct.Queue.html#method.write_buffer_with) that is meant to temporarily lend a view into a write-only buffer, which is similar to part of this discussion. I was wondering if we could do something similar in WebGPU to avoid a copy for common cases where people wouldn't necessarily need to use `mapAsync`/`copyBufferToBuffer` otherwise.

```
// Get an ArrayBuffer backed by memory that will be used for the encoded writeBuffer command sent to the GPU process
const tmp = writeBufferWith(/* ... */);
// Create views/write to them directly without the application needing to copy from somewhere else
(new Float32Array(tmp.buffer))[0] = 123;
// Somehow detach the ArrayBuffer explicitly so the implementation knows we're done writing to it
tmp.finish();
```

**Comment by Jiawei-Shao** [comment]
> A `buffer.mapWriteData(arrayBuffer)` would be interesting but ultimately that's one more copy as well because the application needs to stage the data somewhere and then ask the browser to do the copy.

So `buffer.mapWriteData()` will introduce one more copy in JavaScript, while always setting the return `arraybuffer` of `buffer.getMappedRange()` to 0 will do one more clear. Not sure which one has better performance.

**Comment by kainino0x** [comment]
FYI webgpu.h has already added `wgpuBufferReadMappedRange` and `wgpuBufferWriteMappedRange`, which are are just like getMappedRange+memcpy. We added these because Wasm-JS bindings have similar issues with needing to copy data and clear memory. They allow copying directly from the Wasm heap into JS. The bindings implementation currently uses getMappedRange and then memcpy's between the Wasm heap and the resulting ArrayBuffer. If we had this in JS then we would use it directly, instead.

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2025-04-09#cannot-uploaddownload-to-uma-storage-buffer-without-an-unnecessary-copy-and-unnecessary-memory-use-2388">GPU Web WG 2025-04-09 Atlantic-time</a></summary>

* CW: close to a solution. This issue is difficult. Problem - no way to get the optimal perf/# copies for both impls that can triply map buffers, and impls that can't, for when you do map_write. We don't want to copy data from GPU when we can't triply map, so we zero out - or in triply mapped, we give you the GPU's data because doing so is free and everything else is work. My hunch - we are going to be able to triply map buffers in all impls eventually, so we should optimize for that, but don't want to predict that. Think only 70% of the cases in Chromium can do this.
* GT: can you make those two separate cases? You ask for triply mapped, but if we can't give that to you we don't give you that option?
* CW: might not be possible to know in the JavaScript process whether triple mapping will work.
* GT: more that, it's a device/adapter feature. Triply mapped is something you can ask for at adapter/device.
* CW: not sure this works consistently even on a single device. Could gather that data point. If works, makes things much easier.
* CW: shouldn't expose rebar to JS because of caching issues.
* JB: so you expect 2 classes of GPUs. When they can share memory efficiently, can also get that memory into the content process efficiently.
* CW: yes.
* CW: if we're willing to make that guess, then the UMA buffer mapping spec gets easier. When you map-write, you always get the previous contents of the buffer because it's triply mapped.
* KN: is there a perf penalty for getting that data? Cache flushes, etc.?
* CW: don't think so. On Vulkan, the one that has the most exposed knobs and which runs on GPUs with the most constraints - might need to call invalidate(). Invalidates parts of the CPU cache. It's fine. (Might be wrong.) Just need to do this when GPU execution's finished - not that often.
* KN: would probably just impact the time it takes to map.
* CW: I'm kind of guessing at this point.
* CW: everyone OK if we try going in that direction? Try asking for impl experience before landing in the spec?
* KN: this extension would be gated on triple mapping?
* CW: yes.
* CW: ok if we move forward? Say this'll be the way this works, but we'll get more impl confidence before landing it?
* JB: so people that have a separate shmem, they'll have an overhead of copying, until they implement triple mapping? Or, don't expose the extension unless triple mapping is available.
* CW: yes. I would advocate we not expose the extension unless you have triple mapping.
* KG: earlier we were thinking UMA shouldn't be an extension (?)
* CW: first meeting we discussed this, thought we were comfortable with this being an extension.

</details>
